### PR TITLE
[autoscaler][observability] {pending, active, failed} metrics by node type

### LINKED ISSUE #33550: [autoscaler][observability][dashboard] Wrong metrics are being used for autoscaler graphs in dashboard
### What happened + What you expected to happen

The pending nodes graph in the dashboard is using the wrong metrics. The metrics it uses don't get reset to 0 when pending nodes goes away. The root cause is it's not using the metrics which the autoscaler produces, it's using some strange metrics from reporter_head which no one owns.

### Versions / Dependencies

Master

### Reproduction script

Launch a node via the autoscaler

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR emits metrics for pending, active, and failed nodes, similar to the information in `ray status`, directly from the autoscaler. 

This is the first half of #33550 which will allow the dashboard to use the correct metrics for cluster state instead of the mysterious and buggy metrics currently being displayed.

## Related issue number

Related to but doesn't close #33550 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
cc @gvspraveen

**Comment by scottsun94** [comment]
Is it possible for it to get into 2.4? This is a pretty obvious bug that might hurt users' confidence in our metrics.
@ericl @zhe-thoughts

**Comment by ericl** [comment]
I'd like to understand this part better:

> mysterious and buggy metrics

What was the issue with the previous metrics and how are they being emitted? Have we tried to fix those metrics?

**Comment by wuisawesome** [comment]
There was a (non-public) discussion [here](https://anyscaleteam.slack.com/archives/C03F5UP92EL/p1678825422839299).

The crux of it is that the currently used metrics are being emitted here and have a few problems.

1. The implementation is buggy. (metrics aren't cleared out)
2. The design is bad. (reporter agent has no business parsing autoscaler status reports, and it can't report this metric in a non-buggy way without violating lots of abstractions).
3. There is now owner. The metrics which the autoscaler owns are defined in `autoscaler/_private/prom_metrics.py`.

As a reminder, the conclusion we came to was that we should migrate to metrics maintained by the autoscaler team. We'll audit the metrics, decide the ones we want to make stable, and change the prefix from `autoscaler_` to `ray_`.

**Comment by DmitriGekhtman** [comment]
Main request from my end is to make sure that nothing breaks (besides some missing metrics) if there's no monitor.py process running at all.

**Comment by wuisawesome** [comment]
@DmitriGekhtman i don't think that holds true today, or is something that we want to hold true, i think we need a different way of meeting the underlying requirement.

**Comment by DmitriGekhtman** [comment]
> i don't think that holds true today

Interesting -- seems to work fine in Ray 1.11.0.
If that's not the case when we upgrade Ray versions, we can turn it back on.

**Comment by wuisawesome** [comment]
> Interesting -- seems to work fine in Ray 1.11.0.
I assume you're missing prometheus metrics, autoscaler event summaries, `ray status` doesn't work, etc.The missing prometheus metrics are just going to hurt more now since they're getting displayed in the dashboard. (let's have this discussion somewhere else though?

**Comment by DmitriGekhtman** [comment]
> prometheus metrics, autoscaler event summaries, ray status doesn't work

None of this is useful to us, since we don't use the autoscaler, have a generic infra observability setup, and do not interact with running Ray nodes by execing into them (so ray status doesn't make sense as a diagnostic tool).

**Comment by DmitriGekhtman** [comment]
We turned off the autoscaler because it was doing nothing but producing obnoxious status logs.
However, that's been solved here: https://github.com/ray-project/ray/pull/31869.
We also solved it internally by turning off RAY_LOG_TO_STDERR and, with nontrivial effort, adjusting our logging setup to follow the [log-tailing sidecar advice](https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/logging.html).
Which means we can turn the autoscaler back on, if need be.

File: python/ray/autoscaler/_private/monitor.py [hunk 1]
```diff
@@ -453,6 +444,57 @@ def _run(self):
...
+                SessionName=self.prom_metrics.session_name,
+                NodeType=node_type,
+            ).set(count)
+
+        failed_node_counts = Counter()
+        for _, node_type in autoscaler_summary.failed_nodes:
+            failed_node_counts[node_type] += 1
+
+        # NOTE: We don't reset this metric very often, so it's possible that if
+        # nodes don't fail frequently, these values could be sticky. We can
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/monitor.py:489]: Could you make the comment more specific?
How often is the metric reset and what do you mean by stick values?

**Comment by rkooo567** [comment]
Without this PR, the pending_nodes will never clean up, which will look pretty bad in the dashboard (because it is on the landing page). 

However, this PR it not enough to fix the issue. We need one more PR to merge in order to fix the issue.

cc @ericl, if you sign off, I will merge this PR and make another PR to fix the issue. Let me know what you think

**Review by rkooo567 — APPROVED** [review]

**Comment by zhe-thoughts** [comment]
For this one I will delegate to @ericl to approve and merge

**Comment by wuisawesome** [comment]
Tests are passing and all comments are addressed. I'm assuming when y'all are talking about "merging" y'all are talking about cherry-picking into the 2.4.0 release branch. 

I will merge into master for now and leave it up to you y'all how you want to proceed with a potential cherry-pick.

File: python/ray/autoscaler/_private/monitor.py (hunk 1 above)
Comment (wuisawesome) [reply python/ray/autoscaler/_private/monitor.py]: added

**Comment by rkooo567** [comment]
sgtm @wuisawesome

**Review by DmitriGekhtman — APPROVED** [review]

File: python/ray/autoscaler/_private/monitor.py [hunk 2]
```diff
@@ -453,6 +444,57 @@ def _run(self):
...
+            self.prom_metrics.pending_resources.labels(
+                resource=resource_name,
+                SessionName=self.prom_metrics.session_name,
+            ).set(pending)
+
+        pending_node_count = Counter()
+        for _, node_type, _ in autoscaler_summary.pending_nodes:
+            pending_node_count[node_type] += 1
+
+        for node_type, count in autoscaler_summary.pending_launches:
```
Comment (scv119) [inline python/ray/autoscaler/_private/monitor.py:467]: looks this causing the bug 

```
econds to complete the update iteration.
2023-05-26 23:46:17,611 ERROR monitor.py:439 -- Monitor: Execution exception. Trying again...
Traceback (most recent call last):
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/py_deps_autoscaler_nightly_ray/site-packages/ray/autoscaler/_private/monitor.py", line 406, in _run
    self.autoscaler.all_node_types,
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/py_deps_autoscaler_nightly_ray/site-packages/ray/autoscaler/_private/monitor.py", line 467, in emit_metrics
    for node_type, count in autoscaler_summary.pending_launches:
ValueError: too many values to unpack (expected 2)
```

**Pull request merged**